[REFACTOR/#252] 온보딩 일부 구조를 리펙토링 합니다.

MoonsuKang opened the pull request:
## 📌 ISSUE
<!--이슈 번호 및 제목을 적어주세요!-->
closed #252 

## 📄 Work Description
<!--어떤 작업을 했는지 작성해주세요!-->
### **1. Mavericks MVI 도입 및 초기 셋업**
  - Mavericks를 활용하여 Presentation Layer에 MVI 패턴 적용
  - 상태 기반 UI 구현 및 단방향 데이터 흐름 정립
 
### **2. Splash / Login / SignUp 흐름에 MVI 구조 반영**
  - 각각의 화면에 MVI 구조 적용 (ViewModel → State, Event, SideEffect 구조화)

### **3. 기존 screen 분리 구조를 SignUpRoute 단일화로 통합**
  - TermsOfService - NickName을 Page로 묶어 단일 Route로 처리

### **4. 핵심 의존성 추상화 (FcmTokenProvider, LoginSdk)**
  - FcmTokenProvider, LoginSdk 등 핵심 의존성을 인터페이스 기반으로 추상화

### **5. ViewModel 개선 및 SOC 기반 의존성 주입 적용**

### **6. BasePreview 도입**
  - BuildTime 시점에서 여러 기기대응을 할 수 있는 PreView Func 추가함

## ✨ PR Point
<!--리뷰어가 특별히 봐주었으면 하는 부분이 있다면 작성해주세요!! 코드를 넣어도 됩니다!-->
코드 중 이해가 안되는 부분이 있으면 우선 [Maverick 정리](https://organic-nerve-fff.notion.site/Mavericks-MVI-1cd60749382e80c88987ed04b89005d5) 이거 한번 확인해주시면 감사하겠습니다잉

커밋 내용에 추가 설명이 있으니 한 번 확인해보는 것도 좋을지도?"

## 📸 ScreenShot/Video
<!--스크린샷이나 영상을 첨부해주세요! 생략 하셔도 무방합니다!-->


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Summary by CodeRabbit

- **New Features**
  - Introduced Mavericks-based state management for login, signup, and splash screens, enabling more robust and predictable UI flows.
  - Added structured contracts (state, intents, side effects) for login, signup, and splash processes.
  - Implemented new composable screens for signup steps, including terms of service and nickname input, with improved validation and feedback.
  - Added custom annotation and preview setup for consistent UI previews.
  - Integrated Kakao login functionality with comprehensive login, logout, and unlink support.
  - Added Firebase Cloud Messaging token provider for push notification integration.

- **Improvements**
  - Refactored navigation and screen flow for authentication, consolidating signup steps into a single streamlined process.
  - Enhanced error handling and dialog feedback for login and signup screens.
  - Simplified UI components such as nickname input field for better usability and code clarity.
  - Updated theme usage for consistent styling across the app.
  - Improved lifecycle handling for state and side effects, reducing potential UI inconsistencies.
  - Replaced scaffold layout with simpler container for main app UI.

- **Bug Fixes**
  - Addressed error handling in login and signup flows to provide clearer user feedback.

- **Chores**
  - Added Mavericks library dependencies and initialized Mavericks in the application.
  - Updated dependency management and internal code structure for better maintainability.
  - Reorganized package structure and imports for calendar and ad modules.
  - Introduced lifecycle-aware coroutine helper for cleaner state collection.
  - Renamed theme function for consistency.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

MoonsuKang committed:
[ADD/#252] Mavericks 라이브러리 의존성 추가

MoonsuKang committed:
[ADD/#252] Mavericks 초기화 추가

MoonsuKang committed:
[ADD/#252] LifecycleOwner repeatOnStarted 확장 함수 추가

MoonsuKang committed:
[ADD/#252] BasePreview 컴포저블 구현

MoonsuKang committed:
[REFACTOR/#252] Splash에 Mavericks Mvi 구조 적용

MoonsuKang committed:
[REFACTOR/#252] FCM 토큰 로직 core 패키지로 분리하여 FcmTokenProvider로 추출

MoonsuKang committed:
[REFACTOR/#252] 소셜 로그인 로직 core:login 패키지로 분리 및 LoginSdk 추상화

- KakaoLoginSdk 클래스를 별도로 분리해 login/logout/unlink 책임 위임
- ViewModel 내 직접 호출하던 카카오 로그인 로직 제거
- LoginSdk, LoginAccessToken 등 인터페이스 및 value class 도입
- DI 구조(LoginModule) 정리로 테스트와 확장성 개선

MoonsuKang committed:
[REFACTOR/#252] Login에 Mavericks Mvi 구조 및 soc 적용

MoonsuKang committed:
[REFACTOR/#252] SignUp(회원가입) 흐름 Route 단일화 및 MVI 기반 구조 개선

- TermsOfService, Nickname screen 분리 구조를 SignUpRoute 단일화 구조로 통합
- Mavericks 기반 MVI 패턴으로 SignUp 상태 및 이벤트 처리 구조 전환
- fcmTokenProvider, LoginSdk 등 SOC 기반 의존성 분리 적용
- back navigation 대응을 위한 navigateToPrevious 전달 구조 반영

MoonsuKang committed:
[REFACTOR/#252] SignUp 구조 변경에 따른 navigation parameter 수정

MoonsuKang committed:
[REFACTOR/#252] NickNameTextField 구조 개선 및 상태 관리를 TextFieldValue → String 으로 변경

- 기존 TextFieldValue 기반 로직을 단순한 String 기반으로 리팩토링
- 내부 상태를 ViewModel에서 통합 관리할 수 있도록 구조 정비

MoonsuKang committed:
[REFACTOR/#252] SignUp 흐름 통합 및 parameter 수정

MoonsuKang committed:
[ADD/#252] Mavericks ViewModel 바인딩을 위한 ViewModelsModule 추가

MoonsuKang committed:
[REFACTOR/#252] Scaffold를 Box로 변경하여 레이아웃 단순화

MoonsuKang committed:
[DEL/#252] 사용하지 않는 코드 제거

MoonsuKang committed:
[MOD/#252]: CLODYTheme -> ClodyTheme로 수정

MoonsuKang committed:
[REFACTOR/#252] 패키지 구조 변경

- 광고 관련 클래스들을 ad 패키지로 이동
- calendar을 calendar 패키지로 이동

MoonsuKang committed:
trigger CI

MoonsuKang committed:
[FIX/#252] Ktlint formatting

SYAAINN commented on app/src/main/java/com/sopt/clody/presentation/utils/extension/LifeCycle.kt:
```diff
@@ -0,0 +1,27 @@
+ *
+ * @param block STARTED 상태에서 실행할 suspend 함수 블록
+ *
+ * @see Lifecycle.repeatOnLifecycle
+ * @see Lifecycle.State.STARTED
```
[3] 
@see 는 관찰하고 있다 이런 뜻인가요?

MoonsuKang replied:
관련된 API나 문서를 참고하라는 의미로 사용됩니다잉
[Kdoc Ref](https://kotlinlang.org/docs/kotlin-doc.html#see-identifier)

SYAAINN commented on app/src/main/java/com/sopt/clody/presentation/ui/splash/SplashContract.kt:
[3]
타 프로젝트에서 MVI를 사용하면서 고민했던 문제 중 하나인데, Intent와 SideEffect를 구분하는 기준이 궁금합니다!
유저가 클릭하는 등의 인터렉션이 없어도 뷰를 진입하면서 시스템 내부적으로 진행되는 동작들이 있을텐데(ex. 자동로그인 검증 등) 이런 부분에서 유저가 의도하지 않아도 일어나는 것들도 Intent일까? 라는 궁금증이 들었기 때문인데요, 
나름대로 내린 결론은 **상태의 변경 여부** 였습니다. State를 변경하는 일체의 동작들은 Intent로 규정하고, 그 외 동작들은 SideEffect로 규정하였습니다. 어떻게 생각하시나요 😀

MoonsuKang replied:
음.. "상태 변경 여부"를 기준으로 Intent, SideEffect를 구분하는 방법 좋습니다.

MVI에서는 Intent가 State를 변경하기 위한 주요 진입점 중 하나고,(Intent - reduce - state)
요 구조를 통해서 상태 변경의 흐름을 추적하고 예측 가능하게 만드는 것이 이 디자인 패턴의 목적이죠.

Toast, SnackBar, Navigate Func 같이 일회성 이벤트고 state를 남길 필요가 없는 행위들을 SideEffect로 구분하는게 적절하다고 생각합니다.
이런 동작까지 State로 표현하면 오히려 관리성 측면에서 복잡해지고 명확성이 떨어질 수 있으니깐요.

결론적으로, 유저의 '의도' 가 없는 시스템 동작이라도 state를 변화시키는 지의 여부를 중심으로 Intent, SideEffect를 나누는 기준은 매우 좋은 접근 방법이라고 생각합니다ㅎ

SYAAINN commented on app/src/main/java/com/sopt/clody/presentation/ui/splash/SplashScreen.kt:
[3] 
몇 가지 궁금중 남깁니다 ..!
1. LaunchedEffect에서 viewModel을 변수로 받는 이유? Unit으로 해도 충분하지 않을까 생각합니다. ViewModel이 변경되는 일이 거의 없을 것 같기도 하고, 바뀔 때 기존 block을 취소하고 새로 실행되는 과정에서 발생하는 문제점은 없을까요? 또한 SideEffect를 collect 하는 것도 한번 실행해놓으면 자동으로 변경될 때마다 emit을 받아오는 것으로 알고 있어서 viewModel을 구독해야하는 이유가 궁금합니다!
2. 아래와 같이 앱 종료나 마켓 이동 등의 동작도 SideEffect로 관리하는 것이 MVI 적으로 일관성 있지 않을까 생각합니다!
```
// ViewModel에서
_sideEffects.send(SplashContract.SplashSideEffect.NavigateToMarketAndFinish)
_sideEffects.send(SplashContract.SplashSideEffect.FinishApp)

// View에서
when (effect) {
    is NavigateToMarketAndFinish -> AppUpdateUtils.navigateToMarketAndFinish(context as Activity)
    is FinishApp -> (context as Activity).finishAffinity()
}
```

MoonsuKang replied:
1. viewmodel을 key로 설정한 이유는 ViewModel Instance가 교체되는 상황에서도 초기화 로직을 정확히 재실행 하기 위해서 사용했씁니다. 
예를들어서 프로세스 restore, nav pop 같은 경우 mavericksViewmodel()가 new instance 가능성이 있기 때문이죠.
이때 Viewmodel 내부의 state는 초기화가 되는데, 외부에서 실행하던 postIntent나 SideEffect같은 수집 로직이 재실행 안된다면? 어떻게 될까요?(나도 잘 모름..)

Splash처럼 진입점 역할을 하는 View단에서는 혹시나 흐름이 꼬인다면 사용자 경험에 영항이 갈수가 있어서 ViewModel이 재생성 되도 초기 kintent를 정확히 전달하고 , SideEffect 수집이 resumse되는게 중요하다고 생각했습니다.

요러한 사고 과정을 통해 viewModel Instance를 key로 사용해서 viewModel이 바뀌는 경우(configuration change, death & recreation 등등)에도 LauchedEffect가 재실행되게 명시했습니다.

2. 오호 좋습니다잉 관리하는 측면이 훨씬 좋아보이네요

SYAAINN commented on app/src/main/java/com/sopt/clody/presentation/ui/splash/SplashViewModel.kt:
[3]
Factory 의존성 부분이 너무 어려워옹.. ㅠㅠㅠㅠ 
뷰모델에서 공통되게 사용될 부분들을 템플릿화 시켜서 BaseMavericksViewModel을 만들어보는 것은 어떨까요?

SYAAINN commented on app/src/main/java/com/sopt/clody/presentation/ui/login/LoginViewModel.kt:
```diff
@@ -0,0 +1,101 @@
+    private suspend fun handleIntent(intent: LoginContract.LoginIntent) {
+        when (intent) {
+            is LoginContract.LoginIntent.LoginWithKakao -> loginWithKakao(intent.context)
+            LoginContract.LoginIntent.ClearError -> setState { copy(errorMessage = null) }
```
[3]
sealed class라 else문이 불필요하고 마지막에 이렇게 해도 되긴하는군요! 하지만 통일성을 위해 is를 붙이는 것은 어떨까요 ?

SYAAINN commented on app/src/main/java/com/sopt/clody/presentation/ui/auth/signup/SignUpContract.kt:
```diff
@@ -0,0 +1,46 @@
+        val errorMessage: String? = null,
+        val serviceChecked: Boolean = false,
+        val privacyChecked: Boolean = false,
+        val allChecked: Boolean = false,
```
[3] 아래처럼 derivedState 활용하면 조금 더 효율적으로 처리할 수 있지 않을까 슥 건의해봅니다 ㅋ.ㅋ
```
val allChecked: Boolean 
   get() = serviceChecked && privacyChecked
```

SYAAINN reviewed (approved):
후하 .. 레전드 난이도 급상승 
이해 못한게 아직 남아있지만 .. 작업을 위해 일단 어프하고 나중에 찾아가겠습니다 ㅜ

MoonsuKang committed:
[REFACTOR/#252] allChecked 상태를 derived property로 변경하여 불필요한 상태 제거

MoonsuKang committed:
[REFACTOR/#252] ClearError Intent 처리 조건에 is 키워드 추가

MoonsuKang committed:
[REFACTOR/#252] 앱 업데이트 흐름을 Intent 및 SideEffect 기반으로 리팩토링 및 ViewModel 함수분리

- Soft/Hard 업데이트 시 직접 처리하던 market 이동 및 종료 로직을 SideEffect로 분리
- HandleHardUpdate, HandleSoftUpdateConfirm Intent 추가
- NavigateToMarket, NavigateToMarketAndFinish, FinishApp SideEffect 추가
- ViewModel 내부에서 Intent를 처리하도록 구조 개선
- SplashRoute에서 UI 이벤트 발생 시 ViewModel에 Intent 전달하도록 변경

MoonsuKang committed:
[REFACTOR/#252] allChecked가 derived property가 됨에 따른 코드 수정

MoonsuKang committed:
[REFACTOR/#252] handleIntent 함수 분리 진행
